Add missing gettext-call to timesince_last_update-tag

th3hamm0r opened the pull request:
Used in the status panel of a page for example:
![wagtail_missing_translation_timesince_last_update](https://user-images.githubusercontent.com/468164/192997630-042b00f7-306e-49d2-8f6e-6d93d7333205.png)

**TBD:** I've also moved the leading space outside the gettext-call.
Maybe I oversee something obvious, maybe there are languages where this spaces are replaced with something else, so they have to be kept, then of course, I will change my PR.
But during the time I'm actively translating for wagtail, I've fixed multiple strings where the first translator missed the white space (which is quite easy, since you don't see them and translators probably often just write their own version instead of copying the original string, if it doesn't contain any placeholders).
For example: https://github.com/wagtail/wagtail-localize/commit/e23dc8cea862016fee9e65c6e2b20f9e061843fa#diff-d34c885092c6e928532288ec254e9e991b4cb6b40f341ca72b5842854c18f3cbR107

If this approach is ok, we may add a point to the i18n discussion #8017 for future strings, @lb- ?
Should we change existing strings with white spaces too, or just keep that in mind for future strings?

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

lb- commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -722,7 +722,11 @@ def timesince_last_update(
+        if user_display_name:
+            by_user = " " + _("by %(user_display_name)s") % {
+                "user_display_name": user_display_name
+            }
```
Here's my thoughts - what if we avoid the clunky `by_user` completely here. Instead let's allow the translator to determine the sentence structure of the time by user and it's spaces. 

If we assume the time is at the start of this label we are probably making a bad assumption, same with assuming there is the 'by' before the name of the thing. 

Instead we should just skip the whole problem by generating a new translation with the sentence of the two parts. The time and the user. 

However, I agree with the concept of trailing/ leading whitespace should not be something that the translator needs to thin about. But usually that means we are doing something odd by building bits of a sentence standalone - which is probably the more critical problem. 

Here is a suggestion. 

Let me know and you think. 


```python
        if user_display_name:
           return _("%(time) by %(user_display_name)s") % {"time": f"{time_prefix}{time_str}", "user_display_name": user_display_name}

        return f"{time_prefix}{time_str}"
```

lb- replied:
The time variable should probably have a better name. Like `time_since`? So it's easier to understand that it's a time ago type time.

th3hamm0r replied:
I like your suggestion, and I thought about storing `f"{time_prefix}{time_str}"` in a new `time_since` variable and use it either in the gettext call or for the simple `return`.
But first I wanted to search for the usage of `time_prefix`, since I couldn't understand what's used there.
I found that there are two places in the templates where such a "time_prefix" is used ([here](https://github.com/wagtail/wagtail/blob/main/wagtail/contrib/modeladmin/templates/modeladmin/includes/header_with_history.html#L11) and [here](https://github.com/wagtail/wagtail/blob/main/wagtail/snippets/templates/wagtailsnippets/snippets/_header_with_history.html#L11)), and in both cases it is the untranslated string "at". I think we have to solve this problem too.

Maybe we replace "time_prefix" with a boolean flag (name?) and if set, we also use a single translation string, so this will result in additional gettext strings. Boolean, because I've only found the usage of the prefix "at". What do you think?

lb- replied:
yeah good call - passing in a single word like that is never really going to work well.

I like your approach. 

Please update accordingly and ping me on the PR and I can take another look. 

Thanks for this.

th3hamm0r replied:
Sry @lb-, there is one more thing :sweat_smile: 
After looking at `timesince_last_update()` again, I think there is another issue:
Currently the `user_display_name` is only used, if the passed date is from the same day, otherwise only the short version is returned. `user_display_name` is used quite often and I don't think that it is really intended, that the username is only shown when the date is from the current day.

If you're ok with that, I would first improve the tests to also contain cases where `user_display_name` is set and the date is older than one day (those would fail atm), then I would remove the calls to `timesince_simple()` from within `timesince_last_update()` and replace them all with separate translation strings, based on the given parameters.

th3hamm0r replied:
Maybe we move this issue to the 4.1 release to avoid blocking 4.0.3? These changes will create new translation strings, so it depends on the translations from transifex, and it is not really a blocker(?)

th3hamm0r replied:
I don't know if I really understand your last sentence, but since this issue has been moved to 4.1, I would now tackle all the i18n issues with `timesince_last_update()` as described in https://github.com/wagtail/wagtail/pull/9261#discussion_r985674134 and update this PR accordingly. I hope that this is also ok for you?

lb- replied:
Thanks. Sorry for the confusing last sentence above. 

I mean that if we changed things for 4.0 branch that created/removed translations and then did it again for 4.1 it would be frustrating for translators. 

However, now that this will be in 4.1 - it's not an issue.

th3hamm0r commented:
I've now replaced the "time_prefix"-param with a "show_time_prefix" boolean flag and implemented separate gettext calls for all variations.
This version also adds the `user_display_name` when the date is older than one day (in the previous version it has been ignored).
I've also removed the initial capital letter of "Just now" in `timesince_simple` and used `capfirst` in the template (which should be conform to what we've discussed in #9219), so this is now in sync with the `timesince_last_update` tag.

th3hamm0r commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -699,19 +699,20 @@ def timesince_simple(d):
 @register.simple_tag
 def timesince_last_update(
-    last_update, time_prefix="", user_display_name="", use_shorthand=True
+    last_update, show_time_prefix=False, user_display_name="", use_shorthand=True
```
- naming ok?
- position ok?

lb- replied:
We could use `show_preposition` but that may be harder to understand haha.

I think this is fine - based on the legacy usage, I think the arg position is fine also.

Thanks for calling this out for a double check.

lb- reviewed (approved):
This is working well for me, thanks for digging deeper and getting to the root cause of this.

I think the translation usage may be a bit verbose but makes it work better for translators.

I might add a comment to that code and explain and will merge in soon.

lb- commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -716,15 +717,43 @@ def timesince_last_update(
-
+        if show_time_prefix:
+            if user_display_name:
+                return _("at %(time)s by %(user_display_name)s") % {
```
Not critical for this PR now, but do you think this area of code would be a good candidate for translators comments? (example below)

I do not do translations myself so maybe these are clear enough as is but thinking it might be a good place to try this out and get feedback from translators.

https://docs.djangoproject.com/en/4.1/topics/i18n/translation/#comments-for-translators

```suggestion
                # Translators: used when time is today
                return _("at %(time)s by %(user_display_name)s") % {
```

th3hamm0r replied:
Sry, I'm too late: those comments could be helpful, in that specific case it probably doesn't really change anything, because the string, using "time" as a placeholder, already speaks for itself.
So merging this PR without that change is ok for me, thanks btw! :smile:

th3hamm0r replied:
Looking at it again, what may help as a comment is, that the versions with the prefix are normally appended to other strings. Personally, if I see such a string, I normally look into the code and check its usages to get an overview.

lb- replied:
Yeah - it may be overkill to use the translator comments. We don't use them anywhere else... Yet. 

Just thought it might be a good use case to try them out and get feedback from translators. 

But again. Not critical and this has been merged and is fine without them. 

Thanks for digging into all this again.
